Support distributed read for other acquisition files

tristpinsm opened the pull request.

tristpinsm commented:
the other part of the unicode index map fix is in radiocosmology/caput#201

jrs65 commented:
Ah interesting. I was facing similar issues with the HFB code. In the end I opted to go a slightly different route and write a draco style container that could read the archived HFB code (chime-experiment/ch_pipeline#99). I figured I could probably do the same for many of the other formats to (e.g. CorrData).

I'll have a look at this later. It is a massive pain to need to be maintaining multiple implementations for distributed IO, so we should see if there are reasons to keep it separate, and if we can combine any of it together.

tristpinsm commented:
> Ah interesting. I was facing similar issues with the HFB code. In the end I opted to go a slightly different route and write a draco style container that could read the archived HFB code ([chime-experiment/ch_pipeline#99](https://github.com/chime-experiment/ch_pipeline/pull/99)). I figured I could probably do the same for many of the other formats to (e.g. CorrData).

Yeah I thought about taking this approach, but decided I didn't want to tackle changing the whole framework, especially because the `CorrData` code is so complicated. In the end doing it this way took more effort than I anticipated anyways...

> I'll have a look at this later. It is a massive pain to need to be maintaining multiple implementations for distributed IO, so we should see if there are reasons to keep it separate, and if we can combine any of it together.

I agree that it makes no sense the way the code is organised at the moment, and it would be much better if everything used a common approach. It might be painful to try and move things over seamlessly and support all the old formats, but just making containers for the current acquisition formats would be an improvement.

jrs65 commented on ch_util/andata.py:
```diff
@@ -296,13 +312,37 @@ def from_acq_h5(
 
         """
 
+        if distributed:
```
I kind of think classes should need to opt in to distributed support. See comments below for why,

jrs65 commented on ch_util/andata.py:
```diff
@@ -296,13 +312,37 @@ def from_acq_h5(
+        datasets=None,
+        out_group=None,
+        **kwargs,
+    ):
```
I think for the sake of people maintaining the module there should be a bit more clarity about what these various base methods are for. One thing that would help would be calling this method something else, e.g. `_from_acq_h5_single`, or somesuch that makes it clearer how this differs. It might also be good to put some extensive comments explaining the relationship between everything.

jrs65 commented on ch_util/andata.py:
```diff
@@ -334,6 +374,144 @@ def from_acq_h5(
+            comm = MPI.COMM_WORLD
+
+        # Determine the total number of frequencies
+        nfreq = None
```
This seems problematic as not all datasets have a frequency axis, e.g. doing `WeatherData.from_acq_h5(..., distributed=True)` and I think this would crash on you complaining about the absence of a frequency axis.

jrs65 replied:
One option would be to allow subclasses to name an axis to distributed the read over. If it's present this routine uses it, if it's not present then you can't do a distributed read and it exits elegantly.

tristpinsm commented:
ok, I think I've addressed your comments @jrs65 (thanks!). It feels like making this work is a hack, so I'm not sure what the right path forward is. I guess getting all of the `CorrData` code integrated with `draco` will not be trivial, but probably moving the other containers in here to that model would be straightforward. And this PR only addresses the latter...

tristpinsm commented:
It's not clear this is the way forward to enable distributed reading of other acquisition files. There is a similar problem with the HFB files, so we can shelf this for now and see how the HFB approach pans out and consider generalising it to all non-visibility acquisition files.
